### ISSUE TITLE
Failing health check but still available

### ISSUE THREAD

**Issue by dirkcjelli** [containous/traefik#5231]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [X] No

### What did you do?
I have a small nginx that I ocassionally take "offline" by deliberately failing the health check, which works with traefik 1.7.12 but no longer in v2.0.

I have configured the health check as below and *see* the health check in the nginx log giving a `404`, but even then, the server is still available. Am I missing something?

### What did you expect to see?

Failing health check takes service offline (works in 1.7.12)

### What did you see instead?

Service still available, `serverStatus` in `/api/rawdata` still `UP`

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.0-beta1
Codename:     faisselle
Go version:   go1.12.7
Built:        2019-07-19T16:04:34Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?
Docker provider is used, labels:
```yaml
    labels:
      - "traefik.docker.network=web"
      - "traefik.enable=true"
      - "traefik.http.routers.stuffhttps.rule=Host(`domain`) || Host(`domain2`)"
      - "traefik.http.routers.stuffhttps.entrypoints=https"
      - "traefik.http.routers.stuffhttps.tls.certresolver=default"
      - "traefik.http.routers.stuffhttps.middlewares=sec@file"

      - "traefik.http.routers.stuffhttp.middlewares=redirect"
      - "traefik.http.routers.stuffhttp.rule=Host(`domain`) || Host(`domain2`)"
      - "traefik.http.routers.stuffhttp.entrypoints=http"
           
      - "traefik.http.middlewares.redirect.headers.sslredirect=true"
      
      - "traefik.http.services.stuff.loadbalancer.healthcheck.path=/health/health"
      - "traefik.http.services.stuff.loadbalancer.healthcheck.interval=15s"
      - "traefik.http.services.stuff.loadbalancer.healthcheck.timeout=1s"
```

nginx log example:

```
172.29.0.3 - - [20/Aug/2019:06:08:14 +0000] "GET /health/health HTTP/1.1" 404 153 "-" "Go-http-client/1.1" "-"                                                                                          
2019/08/20 06:08:29 [error] 6#6: *89 open() "/usr/share/nginx/html/health/health" failed (2: No such file or directory), client: 172.29.0.3, server: localhost, request: "GET /health/health HTTP/1.1", 
host: "172.29.0.6:80"
```

**Comment by richierg** [comment]
Any update on this issue ? 
I'm using traefik 2.0 rc2 and I'm in the same situation.

**Comment by benyanke** [comment]
I am also in the same situation - the logs are noting:

```bash
time="2019-09-13T14:19:08Z" level=warning msg="Health check still failing. Backend: \"backend-service@file\" URL: \"http://backend.server.example.com:8080\" Reason: HTTP request failed: Get http://backend.server.example.com:8080/some/file/script.js: dial tcp 10.20.10.4:8080: connect: connection refused"
```

But the dashboard still shows as healthy (status, green checkmark in services), and requests are still sometimes getting sent to it, resulting in 502 to end users.

**Comment by benyanke** [comment]
This seems like a big deal to release 2.0... is there anything I can do to expedite?

**Comment by benyanke** [comment]
My instance of release 2.0.0. Top two backends are the real servers. The rest are the fake ones, but they're all healthy.

![image](https://user-images.githubusercontent.com/4274911/64982401-bb86cd80-d883-11e9-85cb-e98d092012b3.png)

**Comment by ldez** [comment]
@benyanke to help you can create a minimal reproducible case in a docker-compose file.

**Comment by benyanke** [comment]
will do.

**Comment by benyanke** [comment]
```yml
version: '3.5'

services:

  traefik:
    restart: always
    image: traefik:v2.0
    ports:
      - 80:80
      - 9091:8080
    environment:
      - TRAEFIK_LOG=true
      - TRAEFIK_LOG_FORMAT=common
      - TRAEFIK_LOG_LEVEL=debug
      - TRAEFIK_API=true
      - TRAEFIK_API_DASHBOARD=true
      - TRAEFIK_API_INSECURE=true
      - TRAEFIK_ENTRYPOINTS_HTTP=true
      - TRAEFIK_ENTRYPOINTS_HTTP_ADDRESS=:80
      - TRAEFIK_PROVIDERS_DOCKER=true
      - TRAEFIK_PROVIDERS_DOCKER_NETWORK=backendnet
    labels:
      - "traefik.http.routers.router0.entrypoints=http, traefik"
      - "traefik.http.routers.router0.service=service0"
      - "traefik.http.routers.router0.rule=Host(`127.0.0.1`)"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    links:
      - backtestone:backtestone
      - backtesttwo:backtesttwo
    networks:
      - backend
    depends_on:
      - backtestone
      - backtesttwo


  # This should work, because it's directed to access port 80, where the app is running
  backtestone:
    image: tutum/hello-world:latest
    expose:
      - 80
    networks:
      - backend
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=backendnet"
      - "traefik.http.services.service0.loadbalancer.server.port=80"
      - "traefik.http.services.service0.loadbalancer.healthcheck.interval=5s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.timeout=3s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.port=80"
      - "traefik.http.services.service0.loadbalancer.healthcheck.path=/"

  # This should fail, because it's directed to access port 81
  backtesttwo:
    image: tutum/hello-world:latest
    expose:
      - 80
    networks:
      - backend
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=backendnet"
      - "traefik.http.services.service0.loadbalancer.server.port=81"
      - "traefik.http.services.service0.loadbalancer.healthcheck.interval=5s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.timeout=3s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.port=80"
      - "traefik.http.services.service0.loadbalancer.healthcheck.path=/"


networks:
  backend:
    driver: bridge
    name: backendnet
```

Steps to replicate:

1. `docker-compose up -d`

**Comment by benyanke** [comment]
```yml
version: '3.5'

services:

  traefik:
    restart: always
    image: traefik:v2.0
    ports:
      - 80:80
      - 9091:8080
    environment:
      - TRAEFIK_LOG=true
      - TRAEFIK_LOG_FORMAT=common
      - TRAEFIK_LOG_LEVEL=debug
      - TRAEFIK_API=true
      - TRAEFIK_API_DASHBOARD=true
      - TRAEFIK_API_INSECURE=true
      - TRAEFIK_ENTRYPOINTS_HTTP=true
      - TRAEFIK_ENTRYPOINTS_HTTP_ADDRESS=:80
      - TRAEFIK_ENTRYPOINTS_HTTPALT=true
      - TRAEFIK_ENTRYPOINTS_HTTPALT_ADDRESS=:81
      - TRAEFIK_PROVIDERS_DOCKER=true
      - TRAEFIK_PROVIDERS_DOCKER_NETWORK=backendnet
    labels:
      - "traefik.http.routers.router0.entrypoints=http, traefik"
      - "traefik.http.routers.router0.service=service0"    
      - "traefik.http.routers.router0.rule=Host(`127.0.0.1`)"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    networks:
      - backend
    depends_on:
      - backtestone
      - backtesttwo


  # This should work, because it's directed to access port 80, where the app is running
  backtestone:
    image: tutum/hello-world:latest
    networks:
      - backend
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=backendnet"
      - "traefik.http.services.service0.loadbalancer.server.port=80"
      - "traefik.http.services.service0.loadbalancer.healthcheck.interval=5s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.timeout=3s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.path=/"

  # This should fail, because it's just bash
  backtesttwo:
    image: ubuntu
    command: "bash -c 'while true ; do echo test; sleep 5; done'"
    networks:
      - backend
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=backendnet"
      - "traefik.http.services.service0.loadbalancer.server.port=80"
      - "traefik.http.services.service0.loadbalancer.healthcheck.interval=5s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.timeout=3s"
      - "traefik.http.services.service0.loadbalancer.healthcheck.path=/"



networks:
  backend:
    driver: bridge
    name: backendnet
```


Simple example. First backend app container is listening on 80. Second backend container is just `ubuntu` with a bash loop running. Will not respond to traffic on 80.

Steps to replicate:

1. `docker-compose up -d`
2. wait 5-10 seconds for everything to come up
3. visit `http://localhost:9091/dashboard/#/http/services/service0@docker` and note: both backends are healthy.
![image](https://user-images.githubusercontent.com/4274911/64993958-f9442000-d89c-11e9-8fd5-61b4a8107b99.png)
4. Note in the logs: healthchecks are failing on the second container, because no process is bound to 81.
5. Visit `127.0.0.1`: half of requests will 502, because attempts are being sent to the unhealthy backend.


NOTABLE:
5 is not the case when you remove the second entrypoint, `HTTP_ALT`. The requests are routed properly to only healthy endpoints but the dashboard still shows it as healthy.

**Comment by benyanke** [comment]
In further testing, I have found that step (5) in my comment above only occurs some of the time, but the rest of the steps occur repeatably, perhaps due to a race condition or some other non deterministic behavior.

**Comment by ldez** [comment]
Thanks you! That really help us :+1: 

Some remarks:

- The name of services/routers/... must be unique, so you cannot use the same name for 2 services (`service0`)
- I recommend to not do API routing in insecure mode.

I updated your docker-compose file:

```yml
version: '3.5'

services:

  traefik:
    # restart: always
    image: traefik:v2.0.0-rc4
    ports:
      - 80:80
    command:
      - --log.level=DEBUG
      - --api
      - --entrypoints.web.address=:80
      - --providers.docker.exposedbydefault=false
      # - --providers.docker.network=backendnet
    labels:
      - "traefik.enable=true"
      # - "traefik.docker.network=backendnet"
      - "traefik.http.routers.traefik.rule=Host(`traefik.docker.localhost`)"
      - "traefik.http.routers.traefik.service=api@internal"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    # networks:
    #   - backend

  # This should work, because it's directed to access port 80, where the app is running
  backtestone:
    image: containous/whoami:v1.3.0
    # networks:
    #   - backend
    labels:
      - "traefik.enable=true"
      # - "traefik.docker.network=backendnet"
      - "traefik.http.routers.one.rule=Host(`one.docker.localhost`)"
      - "traefik.http.services.one.loadbalancer.healthcheck.interval=5s"
      - "traefik.http.services.one.loadbalancer.healthcheck.timeout=3s"
      - "traefik.http.services.one.loadbalancer.healthcheck.path=/health"

  # This should fail, because it's just bash
  backtesttwo:
    image: alpine:3.10
    command: "sh -c 'while true ; do echo test; sleep 5; done'"
    # networks:
    #   - backend
    labels:
      - "traefik.enable=true"
      # - "traefik.docker.network=backendnet"
      - "traefik.http.routers.two.rule=Host(`two.docker.localhost`)"
      - "traefik.http.services.two.loadbalancer.server.port=80"
      - "traefik.http.services.two.loadbalancer.healthcheck.interval=5s"
      - "traefik.http.services.two.loadbalancer.healthcheck.timeout=3s"
      - "traefik.http.services.two.loadbalancer.healthcheck.path=/"

# networks:
#   backend:
#     driver: bridge
#     name: backendnet
```

**Comment by benyanke** [comment]
Thanks Idez - I have followed those practices in my production file, I was paring it down for my example.

In the case of the service names, I was trying to simulate a failure of one of the two services - intentionally adding both containers to the same service.


In my production config, it looks like:

```toml
        [[http.services.svc1.loadBalancer.servers]]
          url = "http://backend1.example.com:8000"

        [[http.services.svc1.loadBalancer.servers]]
          url = "http://backend2.example.com:8000"
```

In this case, when one goes offline, the behavior mentioned above occurs. 

Hope that helps explain the purpose of my POC.

**Comment by ldez** [comment]
I understood your goal after posting.

Thanks again.

**Comment by benyanke** [comment]
No worries - it was an odd proof of concept, but the clearest I could come up with.

Thanks for the awesome project. I use it daily in my job, and my personal life.

**Comment by traefiker** [comment]
Closed by #5407.

**Comment by benyanke** [comment]
Just upgraded my LB cluster and confirmed this is resolved in release 2.0.1.

**Comment by dirkcjelli** [comment]
Still does not work for me using Version 2.0.1; according to my nginx, the health check *fails* (with a 404) and the icon in the traefik-dashboard for the server turns red (and is green when the health check succeeds).

But even if the health check fails and the icon is red in the dashboard, I'm still able to download files

**Comment by dduportal** [comment]
Hi @dirkcjelli , could you provide a reproduction case please with the full configuration for traefik and an nginx please?

**Comment by dirkcjelli** [comment]
Hi @dduportal, sure:
**traefik.toml:**
```toml
[providers]
  [providers.file]
      directory = "/etc/traefik/config.d"

[entryPoints]
  [entryPoints.http-internal]
  address = ":10000"
  [entryPoints.traefik]
  address = ":9999"
  [entryPoints.http]
  address = ":80"
  [entryPoints.https]
  address = ":443"

[providers.docker]
  endpoint = "unix:///var/run/docker.sock"
  exposedByDefault = false
  network = "web"

[certificatesResolvers.default.acme]
    email = "my@email"
    storage = "/etc/traefik/acme2.json"
  [certificatesResolvers.default.acme.httpChallenge]
    entryPoint = "http"

[api]
dashboard = true
```
I'm **omitting** my Middleware-config for response headers here (in `/etc/traefik/config.d`) 

**traefik-docker.compose.yaml**
```
version: '2.4'

services:
  proxy:
    cpus: 0.8
    mem_limit: 256M
    image: traefik:v2.0.1
    ports:
      - 80:80
      - 443:443
      - 9999:9999
      - 10000:10000
    networks:
      - web
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /share/d_traefik/etc2:/etc/traefik

networks:
  web:
    external: true
```

**nginx-docker.compose.yaml:**
```
version: "2.4"
networks:
  web:
    external: true
services:
  app:
    cpus: 0.2
    mem_limit: 256M
    restart: always
    image: nginx
    expose:
     - "80"
    labels:
      - "traefik.docker.network=web"
      - "traefik.enable=true"
      - "traefik.http.routers.stuffhttps.rule=Host(`host1`) || Host(`host2`)"
      - "traefik.http.routers.stuffhttps.entrypoints=https"
      - "traefik.http.routers.stuffhttps.tls.certresolver=default"
      - "traefik.http.routers.stuffhttps.middlewares=sec@file"

      - "traefik.http.routers.stuffhttp.middlewares=redirect"
      - "traefik.http.routers.stuffhttp.rule=Host(`host1`) || Host(`host2`)"
      - "traefik.http.routers.stuffhttp.entrypoints=http"
           
      - "traefik.http.middlewares.redirect.headers.sslredirect=true"
      
      - "traefik.http.services.stuffhttps.loadbalancer.healthcheck.path=/health.on"
      - "traefik.http.services.stuffhttps.loadbalancer.healthcheck.interval=15s"
      - "traefik.http.services.stuffhttps.loadbalancer.healthcheck.timeout=2s"

    volumes:
      - /share/d_configs/nginx:/usr/share/nginx/html

    networks:
      - web
```

If you need _**any**_ addiitonal information or more details, just ask; I'll do my best.

**Comment by dduportal** [comment]
Reproducted on a simpler setup (see below). The case arise when there are 2 routers with a common service on a given container. If there are only 1 router and 1 service, then it works at expected.

```yml
version: '3'

services:
  proxy:
    image: traefik:v2.0.1
    command:
      - --entrypoints.web.address=:80
      - --entrypoints.websecure.address=:443
      - --providers.docker
      - --api.insecure
    ports:
      - "80:80"
      - "443:443"
      - "8080:8080"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    labels:
      - "traefik.enable=false"

  app:
    image: nginx:alpine
    labels:
      - "traefik.http.routers.webapphttpsec.rule=Host(`localhost`)"
      - "traefik.http.routers.webapphttpsec.entrypoints=websecure"
      - "traefik.http.routers.webapphttpsec.tls=true"
      - "traefik.http.services.webapphttpsec-svc.loadbalancer.healthcheck.path=/health.on"

      - "traefik.http.routers.webapphttp.rule=Host(`localhost`)"
      - "traefik.http.routers.webapphttp.entrypoints=web"
```

```shell
$ docker-compose up -d
$ docker-compose logs app # Check that there is a 404 for the healthcheck from backend's logs
...

$ curl --insecure --location --verbose --silent https://localhost/index.html # Expected: "HTTP/2 503", but it is "HTTP/2 200".
...
< HTTP/2 200 
...
```


View from the dashboard:

<img width="1193" alt="Screenshot 2019-10-07 at 14 09 45" src="https://user-images.githubusercontent.com/1522731/66310645-09d03f00-e90d-11e9-9d00-fc82c5c1902a.png">

If the router `webapphttp` is commented out, then you have a `HTTP/2 503 - Service Unavailable`  as expected:

```shell
$ #Comment out the 2 labels related to `traefik.http.routers.webapphttp.*` router
$ docker-compose up -d
$ curl --insecure --location --verbose --silent https://localhost/index.html # Expected: "HTTP/2 503", got "HTTP/2 200".
...
< HTTP/2 503 
...
Service Unavailable
$
```

**Comment by dirkcjelli** [comment]
@dduportal Nice find :) 

As I understood this far, my setup above with the 2 distinct routers for http and https  is currently the only one supported when I need a redirection http -> https, right?

**Comment by adam8797** [comment]
I’ve just run into this same issue on v2.0.1 and removing the extra routers (so its 1 router to 1 service) solved the problem. Sounds like @dduportal is right

**Comment by dduportal** [comment]
@dirkcjelli @adam8797 , I confirm that the workaround of having 2 services and 2 routers for the container, and map them 1:1, is working (while waiting for a proper fix).

See the following example, with 1 service for http and the associated router, and 1 service + router for https.
Please note that the healthcheck is only defined for the https service, to be sure there is no confusion: you should have `HTTP/200` on `http://localhost` but `HTTP/503` with https://localhost`. (Of course, if you also define an healthcheck for the http service, it will be `HTTP/503` for both URLS).

```yml
version: '3'

services:
  proxy:
    image: traefik:v2.0.1
    command:
      - --entrypoints.web.address=:80
      - --entrypoints.websecure.address=:443
      - --providers.docker
      - --api.insecure
    ports:
      - "80:80"
      - "443:443"
      - "8080:8080"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    labels:
      - "traefik.enable=false"

  app:
    image: nginx:alpine
    labels:  
    ## 2 routers and 2 services
      - "traefik.http.routers.webapphttpsec.rule=Host(`localhost`)"
      - "traefik.http.routers.webapphttpsec.entrypoints=websecure"
      - "traefik.http.routers.webapphttpsec.tls=true"
      - "traefik.http.routers.webapphttpsec.service=webapphttpsec-svc"
      - "traefik.http.services.webapphttpsec-svc.loadbalancer.server.port=80"
      - "traefik.http.services.webapphttpsec-svc.loadbalancer.healthcheck.path=/health.on"

      - "traefik.http.routers.webapphttp.rule=Host(`localhost`)"
      - "traefik.http.routers.webapphttp.entrypoints=web"
      - "traefik.http.routers.webapphttp.service=webapphttp-svc"
      - "traefik.http.services.webapphttp-svc.loadbalancer.server.port=80"
```

**Comment by strarsis** [comment]
I notice a similar issue with latest Traefik Dockerhub image:
When more than one container is labeled for a frontend, the health checks for all containers simply fail, despite the container can be reached (using `wget` from within the running Traefik container).
When only one container is labeled for a frontend, the health check is OK and I can connect to it over Traefik.

**Comment by strarsis** [comment]
@ldez: Traefik would be awesome for my development setup. I can access the different docker containers just by a full hostname, plus TLS with self-signed CA. But currently, with this bug I can only have one docker container per frontend, which is only one for localhost port 80.

**Comment by traefiker** [comment]
Closed by #5860.
